appender.go

```diff
@@ -66,214 +136,465 @@ func (a *Appender) Error() error {
-	defaultLogicalType := C.duckdb_create_logical_type(0)
-	rowTypes := C.malloc(C.size_t(len(args)) * C.size_t(unsafe.Sizeof(defaultLogicalType)))
-
-	tmpChunkTypes := (*[1<<30 - 1]C.duckdb_logical_type)(rowTypes)
+// Create an array of DuckDB types from a list of Go types.
+func (a *Appender) initColTypes(args []driver.Value) error {
+	a.colInfos = make([]colInfo, len(args))
+	a.colTypesPtr, a.colTypes = mallocLogicalTypeSlice(len(args))
 
 	for i, val := range args {
-		switch v := val.(type) {
-		case uint8:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_UTINYINT)
-		case int8:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_TINYINT)
-		case uint16:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_USMALLINT)
-		case int16:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_SMALLINT)
-		case uint32:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_UINTEGER)
-		case int32:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_INTEGER)
-		case uint64, uint:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_UBIGINT)
-		case int64, int:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_BIGINT)
-		case float32:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_FLOAT)
-		case float64:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_DOUBLE)
-		case bool:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_BOOLEAN)
-		case []byte:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_BLOB)
-		case UUID:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_UUID)
-		case string:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_VARCHAR)
-		case time.Time:
-			tmpChunkTypes[i] = C.duckdb_create_logical_type(C.DUCKDB_TYPE_TIMESTAMP)
-		default:
-			panic(fmt.Sprintf("couldn't append unsupported parameter %T", v))
+		if val == nil {
+			return fmt.Errorf("the first row cannot contain null values (column %d)", i)
 		}
+
+		v := reflect.ValueOf(val)
+		a.colInfos[i], a.colTypes[i] = a.initColInfos(v.Type(), i)
 	}
 
-	a.chunkTypes = (*C.duckdb_logical_type)(rowTypes)
+	return nil
+}
+
+func mallocLogicalTypeSlice(count int) (unsafe.Pointer, []C.duckdb_logical_type) {
+	var dummy C.duckdb_logical_type
+	size := C.size_t(unsafe.Sizeof(dummy))
+
+	ctPtr := unsafe.Pointer(C.malloc(C.size_t(count) * size))
+	slice := (*[1 << 30]C.duckdb_logical_type)(ctPtr)[:count:count]
+
+	return ctPtr, slice
 }
 
-func (a *Appender) addChunk(colCount int) error {
-	a.currentRow = 0
+func mallocCStringSlice(count int) (unsafe.Pointer, []*C.char) {
+	var dummy *C.char
+	size := C.size_t(unsafe.Sizeof(dummy))
+
+	csPtr := unsafe.Pointer(C.malloc(C.size_t(count) * size))
+	slice := (*[1 << 30]*C.char)(csPtr)[:count:count]
+	return csPtr, slice
+}
+
+func initPrimitive[T any](ddbType C.duckdb_type) (colInfo, C.duckdb_logical_type) {
```

Comment:
Awesome use of generics. Makes the code a lot cleaner 🎉
Maybe we should rename `ddbType` to `duckdbType` here to.